### PR TITLE
LINK-1641 | Audit Log to DB

### PR DESCRIPTION
**Pull request by harriris-vincit**
### Description
Adds audit logging for all events, registrations and GDPR API operations. This is achieved through `AuditLogMiddleware`. The log entries are stored into the DB as `AuditLogEntry` rows where the `message` field contains the log data.
### Closes
[LINK-1641](https://helsinkisolutionoffice.atlassian.net/browse/LINK-1641)

[LINK-1641]: https://helsinkisolutionoffice.atlassian.net/browse/LINK-1641?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: events/tests/test_permissions.py [hunk 1]
```diff
@@ -84,18 +86,22 @@ def test_can_edit_event(membership_status, expected_public, expected_draft):
...
-        (True, False, False),
-        (False, True, False),
-        (False, False, True),
+        (True, False, True, False),
+        (True, False, False, False),
+        (False, False, True, False),
+        (True, True, False, False),
+        (False, True, True, False),
+        (False, True, False, False),
+        (False, False, False, True),
```
Comment (jorilindell) [inline events/tests/test_permissions.py:99]: `(True, True, True, False)` combination is missing here. Not a big dealt. Just mentioning if it's not intentional.

**Review by jorilindell — APPROVED** [review]
Looks very good! I added a minor comment about a possible missing test combination. You can ignore that if it's intentional

File: events/tests/test_permissions.py (hunk 1 above)
Comment (harriris-vincit) [reply events/tests/test_permissions.py:100]: Very good catch! I fixed it in 56b8b1e55b9e97fc48220fa59e80898c51bacabd.

File: audit_log/utils.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+def _get_operation_name(request):
+    return _OPERATION_MAPPING.get(request.method, f"Unknown: {request.method}")
+
+
+def _get_remote_address(request):
+    if not request.headers.get("x-forwarded-for"):
+        return request.META.get("REMOTE_ADDR")
+
+    remote_addr = request.headers.get("x-forwarded-for", "").split(",")[0]
```
[the remark this answers is not in the document]
Comment (harriris-vincit) [reply audit_log/utils.py]: Another good catch which was fixed in 8e99d809e27dc8c81bf12d476b94309acc5689c5.

File: audit_log/utils.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        remote_addr = remote_addr[1:].split("]")[0]
+
+    return remote_addr
+
+
+def _get_user_role(user):
+    if user is None:
+        return Role.SYSTEM.value
+
+    if isinstance(user, AnonymousUser):
```
[the remark this answers is not in the document]
Comment (harriris-vincit) [reply audit_log/utils.py]: Fixed in ca33c00be5dc223b503a2f4f79edd90d9fa5c294

File: audit_log/utils.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+    return {
+        "role": _get_user_role(user),
+        "uuid": str(uuid) if uuid else None,
+        "ip_address": _get_remote_address(request),
+    }
+
+
+def commit_to_audit_log(request, response):
+    current_time = datetime.now(tz=timezone.utc)
```
[the remark this answers is not in the document]
Comment (harriris-vincit) [reply audit_log/utils.py]: Fixed in ca33c00be5dc223b503a2f4f79edd90d9fa5c294

File: audit_log/utils.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    iso_8601_date = f"{current_time.replace(tzinfo=None).isoformat(sep='T', timespec='milliseconds')}Z"
```
[the remark this answers is not in the document]
Comment (harriris-vincit) [reply audit_log/utils.py]: Naming improved in ca33c00be5dc223b503a2f4f79edd90d9fa5c294

**Pull request merged**